src/main/scala/com/wavesplatform/history/HistoryWriterImpl.scala

```diff
@@ -40,29 +42,35 @@ class HistoryWriterImpl private(file: Option[File], val synchronizationToken: Re
 
   private def displayFeatures(s: Set[Short]): String = s"FEATURE${if (s.size > 1) "S"} ${s.mkString(", ")} ${if (s.size > 1) "WERE" else "WAS"}"
 
+  private def featureVotesCountInActivationWindow(windowClosingHeight: Int, ngState: Option[NgState] = None): Map[Short, Int] = read { implicit lock =>
+    (Math.max(1, windowClosingHeight - FeatureApprovalBlocksCount + 1 + ngState.size) to windowClosingHeight)
+      .flatMap(featuresAtHeight().get(_))
+      .foldLeft(ngState.map(ng => ng.base.supportedFeaturesIds.map(_ -> 1).toMap).getOrElse(Map.empty[Short, Int])) { (counts, feature) =>
+        counts.updated(feature, counts.getOrElse(feature, 0) + 1)
+      }
+  }
+
+  private def acceptedFeaturesInActivationWindow(windowClosingHeight: Int, ngState: Option[NgState] = None): Set[Short] = {
+    featureVotesCountInActivationWindow(windowClosingHeight, ngState)
+      .filter(p => p._2 >= MinVotesCountToActivateFeature)
+      .keySet
+  }
+
   private def updateFeaturesState(height: Int): Unit = write { implicit lock =>
     require(height % FeatureApprovalBlocksCount == 0, s"Features' state can't be updated at given height: $height")
 
-    val acceptedFeatures = (Math.max(1, height - FeatureApprovalBlocksCount + 1) to height)
-      .flatMap(h => featuresAtHeight().get(h))
-      .foldLeft(Map.empty[Short, Int]) { (counts, feature) =>
-        counts.updated(feature, counts.getOrElse(feature, 0) + 1)
-      }.filter(p => p._2 >= MinBlocksCountToActivateFeature)
-      .keys.map(k => k -> FeatureStatus.Accepted.status).toMap
+    val acceptedFeatures = acceptedFeaturesInActivationWindow(height)
```

Comment:
`updateFeaturesState` doesn't need a parameter, since you don't update random heights. It's always `height()+1`.